types/wlr_foreign_toplevel_management_v1.c

```diff
@@ -0,0 +1,503 @@
+		wl_list_remove(&toplevel_output->link);
+		free(toplevel_output);
+	} else {
+		// XXX: log an error? crash?
+	}
+}
+
+static void fill_array_from_toplevel_state(struct wl_array *array,
+		uint32_t state) {
+	if (state & WLR_FOREIGN_TOPLEVEL_STATE_MAXIMIZED) {
+		uint32_t *index = wl_array_add(array, sizeof(uint32_t));
+		*index = ZWLR_FOREIGN_TOPLEVEL_HANDLE_V1_STATE_MAXIMIZED;
+	}
+	if (state & WLR_FOREIGN_TOPLEVEL_STATE_MINIMIZED) {
+		uint32_t *index = wl_array_add(array, sizeof(uint32_t));
+		*index = ZWLR_FOREIGN_TOPLEVEL_HANDLE_V1_STATE_MINIMIZED;
+	}
+	if (state & WLR_FOREIGN_TOPLEVEL_STATE_ACTIVATED) {
+		uint32_t *index = wl_array_add(array, sizeof(uint32_t));
+		*index = ZWLR_FOREIGN_TOPLEVEL_HANDLE_V1_STATE_ACTIVATED;
+	}
+}
+
+static void toplevel_send_state(struct wlr_foreign_toplevel_handle_v1 *toplevel) {
+	struct wl_array states;
+	wl_array_init(&states);
+	fill_array_from_toplevel_state(&states, toplevel->state);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &toplevel->resources) {
+		zwlr_foreign_toplevel_handle_v1_send_state(resource, &states);
+	}
+
+	wl_array_release(&states);
+}
+
+void wlr_foreign_toplevel_handle_v1_set_maximized(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel, bool maximized) {
+	if (maximized) {
+		toplevel->state |= WLR_FOREIGN_TOPLEVEL_STATE_MAXIMIZED;
+	} else {
+		toplevel->state &= ~WLR_FOREIGN_TOPLEVEL_STATE_MAXIMIZED;
+	}
+	toplevel_send_state(toplevel);
+}
+
+void wlr_foreign_toplevel_handle_v1_set_minimized(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel, bool minimized) {
+	if (minimized) {
+		toplevel->state |= WLR_FOREIGN_TOPLEVEL_STATE_MINIMIZED;
+	} else {
+		toplevel->state &= ~WLR_FOREIGN_TOPLEVEL_STATE_MINIMIZED;
+	}
+	toplevel_send_state(toplevel);
+}
+
+void wlr_foreign_toplevel_handle_v1_set_activated(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel, bool activated) {
+	if (activated) {
+		toplevel->state |= WLR_FOREIGN_TOPLEVEL_STATE_ACTIVATED;
+	} else {
+		toplevel->state &= ~WLR_FOREIGN_TOPLEVEL_STATE_ACTIVATED;
+	}
+	toplevel_send_state(toplevel);
+}
+
+void wlr_foreign_toplevel_handle_v1_send_done(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel) {
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &toplevel->resources) {
+		zwlr_foreign_toplevel_handle_v1_send_done(resource);
+	}
+}
+
+void wlr_foreign_toplevel_handle_v1_destroy(
+		struct wlr_foreign_toplevel_handle_v1 *toplevel) {
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &toplevel->resources) {
+		zwlr_foreign_toplevel_handle_v1_send_closed(resource);
+		wl_resource_set_user_data(resource, NULL);
+	}
+
+	wl_list_remove(&toplevel->link);
+	if (toplevel->title) {
+		free(toplevel->title);
+	}
+	if (toplevel->app_id) {
+		free(toplevel->app_id);
+	}
+
+	free(toplevel);
+}
+
+static void foreign_toplevel_resource_destroy(struct wl_resource *resource) {
+	if (wl_resource_get_user_data(resource)) {
```

Comment:
No need to check for this. Just ensure to call `wl_list_init` after removing from the list.